Angular compilation errors when using Module Federation with nested tsconfig paths

### Current Behavior

Using a tsconfig path that is nested within another path with Module Federation in an Angular application will error.

Without Module Federation being used, the application will compile without issues.

Here is an example of a nested tsconfig path:
```json
"paths": {
      "@acme/shared": ["shared/src/index.ts"],
      "@acme/shared/components": ["shared/components/src/index.ts"],
    }
```

### Expected Behavior

It should compile successfully.

### GitHub Repo

https://github.com/gremlin896/mf-nested-paths-issue-repro

### Steps to Reproduce

1. `nx serve host` or `nx build host`


### Nx Report

```shell
Node   : 20.9.0
   OS     : darwin-arm64
   npm    : 10.1.0
   
   nx                 : 17.1.2
   @nx/js             : 17.1.2
   @nx/jest           : 17.1.2
   @nx/linter         : 17.1.2
   @nx/eslint         : 17.1.2
   @nx/workspace      : 17.1.2
   @nx/angular        : 17.1.2
   @nx/cypress        : 17.1.2
   @nx/devkit         : 17.1.2
   @nx/eslint-plugin  : 17.1.2
   @nrwl/tao          : 17.1.2
   @nx/web            : 17.1.2
   @nx/webpack        : 17.1.2
   typescript         : 5.2.2
```


### Failure Logs

```shell
> nx run host:serve:development


>  NX  Starting module federation dev-server for host with 3 remotes


>  NX  Building 3 static remotes...

- Generating browser application bundles (phase: setup)...

- Generating browser application bundles (phase: setup)...

    TypeScript compiler options "target" and "useDefineForClassFields" are set to "ES2022" and "false" respectively by the Angular CLI. To control ECMA version and features use the Browerslist configuration. For more information, see https://angular.io/guide/build#configuring-browser-compatibility

    NOTE: You can set the "target" to "ES2022" in the project's tsconfig to remove this warning.

    TypeScript compiler options "target" and "useDefineForClassFields" are set to "ES2022" and "false" respectively by the Angular CLI. To control ECMA version and features use the Browerslist configuration. For more information, see https://angular.io/guide/build#configuring-browser-compatibility

    NOTE: You can set the "target" to "ES2022" in the project's tsconfig to remove this warning.

- Generating browser application bundles (phase: setup)...

    TypeScript compiler options "target" and "useDefineForClassFields" are set to "ES2022" and "false" respectively by the Angular CLI. To control ECMA version and features use the Browerslist configuration. For more information, see https://angular.io/guide/build#configuring-browser-compatibility

    NOTE: You can set the "target" to "ES2022" in the project's tsconfig to remove this warning.

✔ Browser application bundle generation complete.

✔ Browser application bundle generation complete.

✔ Browser application bundle generation complete.

✔ Browser application bundle generation complete.

- Copying assets...

✔ Browser application bundle generation complete.

- Copying assets...

✔ Browser application bundle generation complete.

- Copying assets...

✔ Copying assets complete.

- Generating index html...

✔ Copying assets complete.

- Generating index html...

✔ Copying assets complete.
- Generating index html...

✔ Index html generation complete.

✔ Index html generation complete.

✔ Index html generation complete.


>  NX  Built 3 static remotes

⠋ Generating browser application bundles (phase: setup)...    TypeScript compiler options "target" and "useDefineForClassFields" are set to "ES2022" and "false" respectively by the Angular CLI. To control ECMA version and features use the Browerslist configuration. For more information, see https://angular.io/guide/build#configuring-browser-compatibility
    NOTE: You can set the "target" to "ES2022" in the project's tsconfig to remove this warning.
✔ Browser application bundle generation complete.

Initial Chunk Files                                                                                     | Names         |  Raw Size
polyfills.js                                                                                            | polyfills     | 364.22 kB | 
styles.css, styles.js                                                                                   | styles        | 261.63 kB | 
vendor.js                                                                                               | vendor        | 228.83 kB | 
main.js                                                                                                 | main          |  39.46 kB | 

                                                                                                        | Initial Total | 894.15 kB

Lazy Chunk Files                                                                                        | Names         |  Raw Size
node_modules_angular_core_fesm2022_core_mjs.js                                                          | -             |   1.37 MB | 
default-node_modules_angular_router_fesm2022_router_mjs.js                                              | -             | 345.51 kB | 
default-node_modules_angular_common_fesm2022_common_mjs.js                                              | -             | 292.13 kB | 
default-node_modules_angular_common_fesm2022_http_mjs.js                                                | -             | 122.13 kB | 
default-node_modules_angular_platform-browser_fesm2022_platform-browser_mjs.js                          | -             |  90.56 kB | 
default-node_modules_rxjs_dist_esm_internal_observable_of_js-node_modules_rxjs_dist_esm_inter-56567f.js | -             |  88.66 kB | 
host_src_bootstrap_ts.js                                                                                | bootstrap     |  46.79 kB | 
node_modules_angular_core_fesm2022_primitives_signals_mjs.js                                            | -             |  21.54 kB | 
common.js                                                                                               | common        |   2.77 kB | 
 | -             |   0 bytes | 
 | -             |   0 bytes | 
 | -             |   0 bytes | 
 | -             |   0 bytes | 
 | -             |   0 bytes | 
 | -             |   0 bytes | 
 | -             |   0 bytes | 
 | -             |   0 bytes | 
 | -             |   0 bytes | 

Build at: 2023-11-16T17:05:29.287Z - Hash: 3e0a4d2f9a377995 - Time: 1314ms

Error: Module not found: Error: Can't resolve '@acme/shared/components' in '/acme/host/src/app'



** Angular Live Development Server is listening on localhost:4200, open your browser on http://localhost:4200/ **


✖ Failed to compile.
Watchpack Error (initial scan): Error: ENOTDIR: not a directory, scandir '/acme/shared/src/index.ts'
```


### Package Manager Version

_No response_

### Operating System

- [X] macOS
- [ ] Linux
- [ ] Windows
- [ ] Other (Please specify)

### Additional Information

The repro repository was created like so:

`npx create-nx-workspace acme --preset=apps`
`nx g @nx/angular:host host --remotes=shop,cart,about`
`nx g @nx/angular:library --name=shared --projectNameAndRootFormat=as-provided`
`nx g @nx/angular:library --name=components --directory=shared --projectNameAndRootFormat=derived`

Following this, the generated components from the two libraries were imported to the `AppComponent` of `host` and used in the template.

--- a/packages/webpack/src/utils/module-federation/share.spec.ts
+++ b/packages/webpack/src/utils/module-federation/share.spec.ts
@@ -58,6 +58,30 @@ describe('MF Share Utils', () => {
       });
     });
 
+    it('should order nested projects first', () => {
+      // ARRANGE
+      jest.spyOn(fs, 'existsSync').mockReturnValue(true);
+      jest.spyOn(tsUtils, 'readTsPathMappings').mockReturnValue({
+        '@myorg/shared': ['/libs/shared/src/index.ts'],
+        '@myorg/shared/components': ['/libs/shared/components/src/index.ts'],
+      });
+
+      // ACT
+      const sharedLibraries = shareWorkspaceLibraries([
+        { name: 'shared', root: 'libs/shared', importKey: '@myorg/shared' },
+        {
+          name: 'shared-components',
+          root: 'libs/shared/components',
+          importKey: '@myorg/shared/components',
+        },
+      ]);
+
+      // ASSERT
+      expect(Object.keys(sharedLibraries.getAliases())[0]).toEqual(
+        '@myorg/shared/components'
+      );
+    });
+
     it('should handle path mappings with wildcards correctly in non-buildable libraries', () => {
       // ARRANGE
       jest.spyOn(fs, 'existsSync').mockImplementation((file: string) => true);

--- a/packages/webpack/src/utils/module-federation/share.ts
+++ b/packages/webpack/src/utils/module-federation/share.ts
@@ -41,8 +41,18 @@ export function shareWorkspaceLibraries(
     return getEmptySharedLibrariesConfig();
   }
 
+  // Nested projects must come first, sort them as such
+  const sortedTsConfigPathAliases = {};
+  Object.keys(tsconfigPathAliases)
+    .sort((a, b) => {
+      return b.split('/').length - a.split('/').length;
+    })
+    .forEach(
+      (key) => (sortedTsConfigPathAliases[key] = tsconfigPathAliases[key])
+    );
+
   const pathMappings: { name: string; path: string }[] = [];
-  for (const [key, paths] of Object.entries(tsconfigPathAliases)) {
+  for (const [key, paths] of Object.entries(sortedTsConfigPathAliases)) {
     const library = workspaceLibs.find((lib) => lib.importKey === key);
     if (!library) {
       continue;
@@ -52,7 +62,7 @@ export function shareWorkspaceLibraries(
     // It will do nothing for React Projects
     collectWorkspaceLibrarySecondaryEntryPoints(
       library,
-      tsconfigPathAliases
+      sortedTsConfigPathAliases
     ).forEach(({ name, path }) =>
       pathMappings.push({
         name,

